README.md
# html-parser-lexer

Parse html content with [state-lexer](https://github.com/mh-cbon/state-lexer).

## Usage

```go
package main

import (
	"bytes"
	"fmt"
	"github.com/mh-cbon/html-parser-lexer"
	"github.com/mh-cbon/state-lexer"
	"os"
	"text/tabwriter"
)

func main() {
	input := `<html>c o n t e n t</html>`

	b := bytes.NewBufferString(input)
	l := htmlparserlexer.NewHtmlParserLexer(b)

	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', tabwriter.Debug)

	fmt.Fprintf(w, "%v\t %v\t %v\n", "name", "type", "value")
	l.Scan(func(tok lexer.Token) {
		fmt.Fprintf(w, "%v\t %v\t %q\n", htmlparserlexer.TokenName(tok), tok.Type, tok.Value)
	})

	w.Flush()
	//Output:
	// name             | type | value
	// TagOpenToken     | 0    | "<html"
	// TagOpenEndToken  | 3    | ">"
	// TextToken        | 11   | "c o n t e n t"
	// TagCloseToken    | 1    | "</html"
	// TagCloseEndToken | 5    | ">"
}
```


CHANGELOG.md
# Changelog - html-parser-lexer

### 1.0.0

__Changes__

- removed TagOpenStartToken and TagCloseStartToken tokens
  to improve usability of the parsing result.
  For a tag such "<html>" is parsed it now returns 2 tokens :
  TagOpenToken    "<html"
  TagOpenEndToken ">"
  For a tag such "</html>" is parsed it now returns 2 tokens :
  TagCloseToken     "</html"
  TagCloseEndToken  ">"
- fix test for removal of TagOpenStartToken and TagCloseStartToken

__Contributors__

- mh-cbon

Released by mh-cbon, Mon 17 Oct 2016 -
[see the diff](https://github.com/mh-cbon/html-parser-lexer/compare/0.0.1...1.0.0#diff)
______________

### 0.0.1

__Changes__

- init

__Contributors__

- mh-cbon

Released by mh-cbon, Sun 16 Oct 2016 -
[see the diff](https://github.com/mh-cbon/html-parser-lexer/compare/487ec1079a4708e9b6801cebc3d390c2e0e52e84...0.0.1#diff)
______________




html-parser-lexer_test.go
package htmlparserlexer

import (
	"bytes"
	"fmt"
	"github.com/mh-cbon/state-lexer"
	"os"
	"testing"
	"text/tabwriter"
)

func Test_Fixture1(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 1, Value: "</html"},
		lexer.Token{Type: 5, Value: ">"},
	}

	input := `<html></html>`

	runTest(t, expected, input, false)
}

func Test_Fixture2(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 10, Value: "\n"},
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 10, Value: "\n"},
		lexer.Token{Type: 0, Value: "<input"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "type"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 9, Value: "button"},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "disabled"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "value"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 9, Value: "42"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 3, Value: "/>"},
		lexer.Token{Type: 10, Value: "\n"},
		lexer.Token{Type: 1, Value: "</html"},
		lexer.Token{Type: 5, Value: ">"},
		lexer.Token{Type: 10, Value: "\n"},
	}

	input := `
<html>
<input type="button" disabled value=42 />
</html>
`

	runTest(t, expected, input, false)
}

func Test_Fixture3(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 0, Value: "<input"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "type"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 9, Value: "butt\\\"on"},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 3, Value: "/>"},
	}

	input := `<input type="butt\"on" />`

	runTest(t, expected, input, false)
}

func Test_Fixture4(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 12, Value: "<!--"},
		lexer.Token{Type: 13, Value: "comments works too"},
		lexer.Token{Type: 14, Value: "-->"},
	}

	input := `<!--comments works too-->`

	runTest(t, expected, input, false)
}

func Test_Fixture5(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 11, Value: "whatever "},
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 11, Value: "it works ?"},
	}

	input := `whatever <html> it works ?`

	runTest(t, expected, input, false)
}

func Test_Fixture6(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 3, Value: ""},
		lexer.Token{Type: 0, Value: "<div"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 11, Value: "shame this..."},
	}

	input := `<html <div> shame this...`

	runTest(t, expected, input, false)
}

func Test_Fixture7(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 10, Value: "\n\n"},
		lexer.Token{Type: 0, Value: "<esi:include"},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "src"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 9, Value: "http://example.com/1.html"},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "alt"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 9, Value: "http://bak.example.com/2.html"},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 10, Value: " "},
		lexer.Token{Type: 6, Value: "onerror"},
		lexer.Token{Type: 7, Value: "="},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 9, Value: "continue"},
		lexer.Token{Type: 8, Value: "\""},
		lexer.Token{Type: 3, Value: "/>"},
		lexer.Token{Type: 10, Value: "\n\n"},
		lexer.Token{Type: 1, Value: "</html"},
		lexer.Token{Type: 5, Value: ">"},
		lexer.Token{Type: 10, Value: "\n"},
	}

	input := `<html>

<esi:include src="http://example.com/1.html" alt="http://bak.example.com/2.html" onerror="continue"/>

</html>
`

	runTest(t, expected, input, false)
}

func Test_Fixture8(t *testing.T) {

	expected := []lexer.Token{
		lexer.Token{Type: 0, Value: "<html"},
		lexer.Token{Type: 3, Value: ">"},
		lexer.Token{Type: 10, Value: "\n  "},
		lexer.Token{Type: 11, Value: "div>\n"},
		lexer.Token{Type: 1, Value: "</html"},
		lexer.Token{Type: 10, Value: "\n"},
		lexer.Token{Type: 5, Value: ""},
		lexer.Token{Type: 11, Value: "Really broken..did you drink ? ah ah ah\n"},
	}

	input := `<html>
  div>
</html
Really broken..did you drink ? ah ah ah
`

	runTest(t, expected, input, false)
}

func runTest(t *testing.T, expected []lexer.Token, input string, debug bool) {

	b := bytes.NewBufferString(input)
	l := NewHtmlParserLexer(b)

	var tokens []lexer.Token
	l.Scan(func(tok lexer.Token) {
		tokens = append(tokens, tok)
	})

	if debug {
		fmt.Printf("%#v", tokens)
	}

	for i, e := range expected {
		if e.Type != tokens[i].Type {
			t.Errorf("Expected token type %v but got %v", e.Type, tokens[i].Type)
		}
		if e.Value != tokens[i].Value {
			t.Errorf("Expected token value %v but got %v", e.Value, tokens[i].Value)
		}
	}

	if len(expected) != len(tokens) {
		t.Errorf("Incoherent parsing expected %v tokens but got %v", len(expected), len(tokens))
	}
}

func Example_lexer() {
	input := `<html>c o n t e n t</html>`

	b := bytes.NewBufferString(input)
	l := NewHtmlParserLexer(b)

	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', tabwriter.Debug)

	fmt.Fprintf(w, "%v\t %v\t %v\n", "name", "type", "value")
	l.Scan(func(tok lexer.Token) {
		fmt.Fprintf(w, "%v\t %v\t %q\n", TokenName(tok), tok.Type, tok.Value)
	})

	w.Flush()
	//Output:
	// name             | type | value
	// TagOpenToken     | 0    | "<html"
	// TagOpenEndToken  | 3    | ">"
	// TextToken        | 11   | "c o n t e n t"
	// TagCloseToken    | 1    | "</html"
	// TagCloseEndToken | 5    | ">"

}


html-parser-lexer.go
package htmlparserlexer

import (
	// "fmt"
	// "time"
	"github.com/mh-cbon/state-lexer"
	"io"
)

const (
	TagOpenToken  lexer.TokenType = iota // 0
	TagCloseToken                        // 1

	yyy              // 4
	TagOpenEndToken  // 3
	xxx              // 4
	TagCloseEndToken // 5

	TagAttrNameToken  // 6
	TagAttrEqToken    // 7
	TagAttrQuoteToken // 8
	TagAttrValueToken // 9

	WsToken   // 10
	TextToken // 11

	CommentStartToken // 12
	CommentToken      // 13
	CommentEndToken   // 14
)

const (
	NOTEOFRune rune = -2
)

func NewHtmlParserLexer(r io.Reader) *lexer.L {
	return lexer.New(r, TextState)
}

func isWhitespace(ch rune) bool { return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' }
func isQuote(ch rune) bool      { return ch == '"' || ch == '\'' }

func eatSomeWs(l *lexer.L) rune {
	hasWs := false
	r := l.Next()
	for isWhitespace(r) {
		hasWs = true
		if r == lexer.EOFRune {
			l.Emit(WsToken)
			return lexer.EOFRune
		}
		r = l.Next()
	}
	l.Rewind()
	if hasWs {
		l.Emit(WsToken)
	}
	return NOTEOFRune
}

func isATagCloseStart(l *lexer.L) bool {
	r := l.Next()
	if r != '<' {
		l.Rewind()
		return false
	}
	r = l.Next()
	if r != '/' {
		l.Rewind()
		l.Rewind()
		return false
	}
	l.Rewind()
	l.Rewind()
	return true
}

func isATagOpenStart(l *lexer.L) bool {
	r := l.Next()
	if r != '<' {
		l.Rewind()
		return false
	}
	r = l.Next()
	if (r < 'a' || r > 'z') && (r < 'A' || r > 'Z') {
		l.Rewind()
		l.Rewind()
		return false
	}
	l.Rewind()
	l.Rewind()
	return true
}

func isATagCommentStartToken(l *lexer.L) bool {
	r := l.Next()
	if r != '<' {
		l.Rewind()
		return false
	}
	r = l.Next()
	if r != '!' {
		l.Rewind()
		l.Rewind()
		return false
	}
	r = l.Next()
	if r != '-' {
		l.Rewind()
		l.Rewind()
		l.Rewind()
		return false
	}
	r = l.Next()
	if r != '-' {
		l.Rewind()
		return false
	}
	l.Rewind()
	l.Rewind()
	l.Rewind()
	l.Rewind()
	return true
}

func isATagCommentEndToken(l *lexer.L) bool {

	if r := l.Next(); r != '-' {
		l.Rewind()
		return false
	}
	if r := l.Next(); r != '-' {
		l.Rewind()
		l.Rewind()
		return false
	}
	if r := l.Next(); r != '>' {
		l.Rewind()
		l.Rewind()
		l.Rewind()
		return false
	}
	l.Rewind()
	l.Rewind()
	l.Rewind()
	return true
}

func TextState(l *lexer.L) lexer.StateFunc {

	// eat some ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	hasText := false
	r := l.Next()
	for r != '<' {
		if r == lexer.EOFRune {
			if hasText {
				l.Emit(TextToken)
			}
			return nil
		}
		hasText = true
		r = l.Next()
	}
	l.Rewind()
	if hasText {
		l.Emit(TextToken)
	}

	// is it goind to be a </
	if isATagCloseStart(l) {
		return TagCloseState
	}
	// or a <
	if isATagOpenStart(l) {
		return TagStartState
	}
	// or a <!--
	if isATagCommentStartToken(l) {
		return TagCommentStartState
	}
	// dunno... :s :s :s yet!
	return TextState
}

func TagCommentStartState(l *lexer.L) lexer.StateFunc {
	// eat <
	if r := l.Next(); r != '<' {
		panic("ohoh not a comment start")
	}
	// eat !
	if r := l.Next(); r != '!' {
		panic("ohoh not a comment start")
	}
	// eat -
	if r := l.Next(); r != '-' {
		panic("ohoh not a comment start")
	}
	// eat -
	if r := l.Next(); r != '-' {
		panic("ohoh not a comment start")
	}
	l.Emit(CommentStartToken)

	for { // @todo do something if the comment is unclosed. possible ?
		if r := l.Next(); r == lexer.EOFRune {
			return nil
		}
		// <-time.After(time.Millisecond)
		if isATagCommentEndToken(l) {
			break
		}
	}
	l.Emit(CommentToken)

	// eat -
	if r := l.Next(); r != '-' {
		panic("ohoh not a comment end")
	}
	// eat -
	if r := l.Next(); r != '-' {
		panic("ohoh not a comment end")
	}
	// eat >
	if r := l.Next(); r != '>' {
		panic("ohoh not a comment end")
	}
	l.Emit(CommentEndToken)

	return TextState
}

func TagStartState(l *lexer.L) lexer.StateFunc {
	r := l.Next()
	// eat <
	if r != '<' {
		panic("ohoh not a tag start")
	}

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	// eat the tag name
	r = l.Next()
	for (r >= 'a' && r <= 'z') || r == ':' {
		r = l.Next()
	}
	l.Rewind()
	l.Emit(TagOpenToken)

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	// eat >
	r = l.Next()
	if r == '>' {
		l.Emit(TagOpenEndToken)
		return TextState
	}
	l.Rewind()

	//badass html, unclosed html tag name.
	r = l.Next()
	if r == '<' {
		l.Rewind()
		l.Emit(TagOpenEndToken) //empty token!!
		return TagStartState
	}
	l.Rewind()

	// go for tag attributes
	return TagAttrNameState
}

func TagCloseState(l *lexer.L) lexer.StateFunc {
	r := l.Next()
	// eat <
	if r != '<' {
		panic("ohoh not a tag close")
	}
	r = l.Next()
	if r != '/' {
		panic("ohoh not a tag close")
	}

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	// eat the tag name
	r = l.Next()
	for r >= 'a' && r <= 'z' {
		r = l.Next()
	}
	l.Rewind()
	l.Emit(TagCloseToken)

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	// eat >
	r = l.Next()
	if r != '>' {
		l.Rewind()
	}
	l.Emit(TagCloseEndToken)
	return TextState
}

func TagAttrNameState(l *lexer.L) lexer.StateFunc {
	// eat attribute name
	r := l.Next()
	for r != '=' && !isWhitespace(r) {
		r = l.Next()
	}
	l.Rewind()
	l.Emit(TagAttrNameToken)

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	// eat eq sign
	withValue := false
	r = l.Next()
	if r == '=' {
		withValue = true
		l.Emit(TagAttrEqToken)
	} else {
		l.Rewind()
	}

	// // eat in between ws
	if r := eatSomeWs(l); r == lexer.EOFRune {
		return nil
	}

	if withValue {
		// eat tag value quote, if any
		r = l.Next()
		if isQuote(r) {
			quote := r // save quote type
			l.Emit(TagAttrQuoteToken)
			// eat the attribute value until a quote
			var p rune // holds previous read to test for escaped quotes
			for {
				r = l.Next()
				for !isQuote(r) {
					p = r
					r = l.Next()
				}
				if r == quote && p != rune('\\') {
					break
				}
			}
			l.Rewind()
			l.Emit(TagAttrValueToken)
			// eat the quote
			l.Next()
			l.Emit(TagAttrQuoteToken)

		} else {
			// eat the attribute value until a whitespace or a >
			r = l.Next()
			for r != '>' && !isWhitespace(r) {
				r = l.Next()
			}
			l.Rewind()
			l.Emit(TagAttrValueToken)
		}

		// // eat in between ws
		if r := eatSomeWs(l); r == lexer.EOFRune {
			return nil
		}

	}

	// close tag, maybe
	r = l.Next()
	if r == '/' {
		r = l.Next()
	}
	if r == '>' {
		l.Emit(TagOpenEndToken)
		return TextState
	}

	// keep looking for attributes
	return TagAttrNameState
}

// Helper function
func TokenName(tok lexer.Token) string {
	switch tok.Type {
	case TagOpenToken:
		return "TagOpenToken"
	case TagCloseToken:
		return "TagCloseToken"
	case TagOpenEndToken:
		return "TagOpenEndToken"
	case TagCloseEndToken:
		return "TagCloseEndToken"
	case TagAttrNameToken:
		return "TagAttrNameToken"
	case TagAttrEqToken:
		return "TagAttrEqToken"
	case TagAttrQuoteToken:
		return "TagAttrQuoteToken"
	case TagAttrValueToken:
		return "TagAttrValueToken"
	case WsToken:
		return "WsToken"
	case TextToken:
		return "TextToken"
	}
	return "undefined"
}
